aspects/src/main/java/org/springframework/security/authorization/method/aspectj/MethodInvocationAdapter.aj

```diff
@@ -0,0 +1,106 @@
+/*
+ * Copyright 2002-2022 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.authorization.method.aspectj;
+
+import java.lang.reflect.AccessibleObject;
+import java.lang.reflect.Method;
+import java.util.function.Supplier;
+
+import org.aopalliance.intercept.MethodInvocation;
+import org.aspectj.lang.JoinPoint;
+import org.aspectj.lang.ProceedingJoinPoint;
+import org.aspectj.lang.reflect.CodeSignature;
+
+import org.springframework.util.Assert;
+
+final class MethodInvocationAdapter implements MethodInvocation {
+
+	private final ProceedingJoinPoint jp;
+
+	private final Supplier<Object> proceed;
+
+	private final Method method;
+
+	private final Object target;
+
+	MethodInvocationAdapter(JoinPoint jp) {
+		this(jp, null);
+	}
+
+	MethodInvocationAdapter(JoinPoint jp, Supplier<Object> proceed) {
```

Comment:
I'm curious about if/why `Supplier<Object>` is necessary. It looks like the `Suplier<Object>` that is always passed in is `jp.proceed()` which would be the same behavior as if `proceed` is `null`.

Additionally, it seems a bit confusing in what circumstances one would override the `JoinPoint.proceed()` behavior. If they did indeed need to do that, wouldn't the `JoinPoint.proceed()` implementation be altered?

Reply:
This may be due to my unfamiliarity with AspectJ. I lifted this from [`AnnotationSecurityAspect`](https://github.com/spring-projects/spring-security/blob/main/aspects/src/main/java/org/springframework/security/access/intercept/aspectj/aspect/AnnotationSecurityAspect.aj#L75) and [`AspectJMethodSecurityInterceptor`](https://github.com/spring-projects/spring-security/blob/main/core/src/main/java/org/springframework/security/access/intercept/aspectj/AspectJMethodSecurityInterceptor.java#L60).

Reply:
I think if we don't need the supplier we don't need to add it now (especially if it is non-public API)

Reply:
It appears to be intentional by the original authors of `AnnotationSecurityAspect` to use the aspect's `proceed()` method instead of the join point's `proceed()` method. When I change the code to use the join point's `proceed()` method, the intercepted method is not invoked correctly. 

When reading [AspectJ's documentation](https://www.eclipse.org/aspectj/doc/next/progguide/examples-basic.html#examples-joinPoints), the same thing is done, e.g. proceeding is triggered by the aspects `proceed()` method instead of the join point's. Because of that, I've changed `JoinPointMethodInvocation` to just take a `JoinPoint` since it appears that `ProceedingJoinPoint#proceed` in this case is not intended to be invoked.